core/trino-main/src/main/java/io/trino/execution/scheduler/policy/PrioritizeUtilizationExecutionSchedule.java

```diff
@@ -0,0 +1,559 @@
+        unblockStagesWithFullOutputBuffer();
+    }
+
+    @VisibleForTesting
+    DirectedGraph<PlanFragmentId, FragmentsEdge> getFragmentDependency()
+    {
+        return fragmentDependency;
+    }
+
+    @VisibleForTesting
+    Set<StageExecution> getActiveStages()
+    {
+        return activeStages;
+    }
+
+    private void removeCompletedStages()
+    {
+        Set<StageExecution> completedStages = activeStages.stream()
+                .filter(this::isStageCompleted)
+                .collect(toImmutableSet());
+        // remove completed stages outside of Java stream to prevent concurrent modification
+        completedStages.forEach(this::removeCompletedStage);
+    }
+
+    private void removeCompletedStage(StageExecution stage)
+    {
+        // start all stages that depend on completed stage
+        PlanFragmentId fragmentId = stage.getFragment().getId();
+        fragmentDependency.outgoingEdgesOf(fragmentId).stream()
+                .map(FragmentsEdge::getTarget)
+                // filter stages that depend on completed stage only
+                .filter(dependentFragmentId -> fragmentDependency.inDegreeOf(dependentFragmentId) == 1)
+                .map(stagesByFragmentId::get)
+                .forEach(this::selectForExecution);
+        fragmentDependency.removeVertex(fragmentId);
+        fragmentTopology.removeVertex(fragmentId);
+        activeStages.remove(stage);
+    }
+
+    private void unblockStagesWithFullOutputBuffer()
+    {
+        // find stages that are blocked on full task output buffer
+        Set<PlanFragmentId> blockedFragments = activeStages.stream()
+                .filter(StageExecution::isAnyTaskBlocked)
+                .map(stage -> stage.getFragment().getId())
+                .collect(toImmutableSet());
+        // start immediate downstream stages so that data can be consumed
+        blockedFragments.stream()
+                .flatMap(fragmentId -> fragmentTopology.outgoingEdgesOf(fragmentId).stream())
+                .map(FragmentsEdge::getTarget)
+                .map(stagesByFragmentId::get)
+                .forEach(this::selectForExecution);
+    }
+
+    private void selectForExecution(StageExecution stage)
+    {
+        if (fragmentDependency.outDegreeOf(stage.getFragment().getId()) > 0) {
+            // if there are any dependent stages then reschedule when stage is completed
+            stage.addStateChangeListener(state -> {
+                if (isStageCompleted(stage)) {
+                    notifyReschedule();
+                }
+            });
+        }
+        activeStages.add(stage);
+    }
+
+    private void notifyReschedule()
+    {
+        SettableFuture<Void> rescheduleFuture;
+        synchronized (this) {
+            rescheduleFuture = this.rescheduleFuture;
+            this.rescheduleFuture = SettableFuture.create();
+        }
+        // notify listeners outside of the critical section
+        rescheduleFuture.set(null);
```

Comment:
This looks racy. 
We can obtain the newly created `this.reschedulFuture` in `PipelinedDistributedStagesScheduler#schedule` just before old one is completed.
Then scheduler will not be woken up.
Not sure if we can do better while keeping `rescheduleFuture.set()` out of critical section - but maybe you will figure something out :)

Reply:
You are right. I've changed `ExecutionSchedule` to return `StagesScheduleResult` which contains both stages and future